src/ispc/kernels/lanczos3.ispc

```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
+    dst = *pixel_ptr3;
+    return dst;
 }
 
-static inline uint8<4> resample_internal(uniform Image src_image, float<2> uv, uniform float<2> target_size, uniform float<2> inv_target_size, uniform uint8 num_channels) {
-    float<2> center_pixel = uv * target_size;
-    center_pixel.x = frac(center_pixel.x);
-    center_pixel.y = frac(center_pixel.y);
-    float<2> center = uv - (center_pixel - 0.5) * inv_target_size;
-    float<2> offset = (uv - center) * target_size;
-    const float inv_255 = rcp(255.0);
-
-    float<4> col = { 0, 0, 0, 0 };
-    float weight = 0.0;
-
-    for(uniform int x = -3; x < 3; x++){
-        for(uniform int y = -3; y < 3; y++){
-
-            float wx = lanczos3_filter((uniform float)x - offset.x);
-            float wy = lanczos3_filter((uniform float)y - offset.y);
-            float w = wx * wy;
-            float<2> texel_offset = {x, y};
-            float<2> texel_uv = center + texel_offset * inv_target_size;
-            int<2> pixel_coord = uv_to_pixel_id(src_image.size, texel_uv);
-
-            int addr = (pixel_coord.x + pixel_coord.y * src_image.size.x) * num_channels;
-
-            float<4> texel;
-
-            if (num_channels == 3) {
-                texel.x = byte_to_float(src_image.data[addr + 0]) * inv_255;
-                texel.y = byte_to_float(src_image.data[addr + 1]) * inv_255;
-                texel.z = byte_to_float(src_image.data[addr + 2]) * inv_255;
-            } else if (num_channels == 4) {
-                texel.x = byte_to_float(src_image.data[addr + 0]) * inv_255;
-                texel.y = byte_to_float(src_image.data[addr + 1]) * inv_255;
-                texel.z = byte_to_float(src_image.data[addr + 2]) * inv_255;
-                texel.w = byte_to_float(src_image.data[addr + 3]) * inv_255;
-            }
-
-            col += w * texel;
-            weight += w;
+void clean_and_write_3_channels(varying float<3> color, varying uint64 write_address, uniform uint8* varying dst) {
+    // The final color is a sum of numbers that are multiplied by the weights of their respective pixels.
+    // Because of their numbers, floating point precision leads to the final color being potentially outside of the 0-255 range by a slight margin.
+    // This would cause an underflow/overflow, which we avoid with the clamps.
+    uniform uint8* pixel_ptr = dst + write_address;
+    pixel_ptr[0] = clamp(color[0], 0.0f, 255.0f);
+    pixel_ptr[1] = clamp(color[1], 0.0f, 255.0f);
+    pixel_ptr[2] = clamp(color[2], 0.0f, 255.0f);
+}
+
+uint8<4> sample_4_channels(varying uint64 read_address, const uniform uint8* varying src) {
+    const uniform uint8* pixel_ptr = src + read_address;
+    const uniform uint8<4>* pixel_ptr4 = (const uniform uint8<4>*)(pixel_ptr);
+    varying uint8<4> dst = {0, 0, 0, 0};
+    dst = *pixel_ptr4;
+    return dst;
+}
+
+void clean_and_write_4_channels(varying float<4> color, varying uint64 write_address, uniform uint8* varying dst) {
```

Comment:
Here and everywhere else: the pointer (`* varying`) is `varying` by default (each program instance points to _a different_ pixel in the buffer), and the pointed-to type (`uint8`) is `uniform` by default (all programs point within the same buffer/memory).